applications/nrf_desktop/src/modules/usb_state.c

```diff
@@ -804,6 +830,359 @@ static int usb_init_legacy(void)
+			new_state = before_suspend;
+		}
+		break;
+
+	case USBD_MSG_SUSPEND:
+		before_suspend = state;
+		new_state = USB_STATE_SUSPENDED;
+		break;
+
+	case USBD_MSG_RESET:
+		if (state != USB_STATE_DISCONNECTED) {
+			new_state = USB_STATE_POWERED;
+		} else {
+			LOG_WRN("Reset while disconnected");
+		}
+		break;
+
+	case USBD_MSG_UDC_ERROR:
+		LOG_ERR("Non-correctable UDC error message");
+		module_set_state(MODULE_STATE_ERROR);
+		break;
+
+	case USBD_MSG_STACK_ERROR:
+		LOG_ERR("Unrecoverable USB device stack error");
+		module_set_state(MODULE_STATE_ERROR);
+		break;
+
+	case USBD_MSG_CDC_ACM_LINE_CODING:
+	case USBD_MSG_CDC_ACM_CONTROL_LINE_STATE:
+		/* Ignore */
+		break;
+
+	default:
+		LOG_ERR("Unexpected USB device message type: %d", msg->type);
+		break;
+	}
+
+	if (new_state != state) {
+		if ((state == USB_STATE_SUSPENDED) && is_usb_active_next()) {
+			broadcast_usb_state(USB_STATE_ACTIVE);
+		} else {
+			broadcast_usb_state(new_state);
+		}
+		state = new_state;
+	}
+}
+
+static int usb_init_next_register_fs_classes(struct usbd_contex *usbd)
+{
+	int err = 0;
+
+	STRUCT_SECTION_FOREACH_ALTERNATE(usbd_class_fs, usbd_class_node, c_nd) {
+		err = usbd_register_class(usbd, c_nd->c_data->name, USBD_SPEED_FS, 1);
+		if (err) {
+			LOG_ERR("usbd_register_class failed for %s (err: %d)",
+				c_nd->c_data->name, err);
+			return err;
+		}
+	}
+
+	return err;
+}
+
+static int usb_init_next_register_hs_classes(struct usbd_contex *usbd)
+{
+	int err = 0;
+
+	STRUCT_SECTION_FOREACH_ALTERNATE(usbd_class_hs, usbd_class_node, c_nd) {
+		err = usbd_register_class(usbd, c_nd->c_data->name, USBD_SPEED_HS, 1);
+		if (err) {
+			LOG_ERR("usbd_register_class failed for %s (err: %d)",
+				c_nd->c_data->name, err);
+			return err;
+		}
+	}
+
+	return err;
+}
+
+static int usb_init_next_add_configuration(struct usbd_contex *usbd,
+					   const enum usbd_speed speed,
+					   struct usbd_config_node *config)
+{
+	int err;
+
+	err = usbd_add_configuration(usbd, speed, config);
+	if (err) {
+		LOG_ERR("usbd_add_configuration failed (err: %d)", err);
+		return err;
+	}
+
+	if (speed == USBD_SPEED_FS) {
+		err = usb_init_next_register_fs_classes(usbd);
+	} else if (speed == USBD_SPEED_HS) {
+		err = usb_init_next_register_hs_classes(usbd);
+	}
+
+	if (err) {
+		return err;
+	}
+
+	/* Always use class code information from Interface Descriptors */
+	if (IS_ENABLED(CONFIG_USBD_CDC_ACM_CLASS)) {
+		BUILD_ASSERT(!IS_ENABLED(CONFIG_USBD_CDC_ECM_CLASS));
+		BUILD_ASSERT(!IS_ENABLED(CONFIG_USBD_AUDIO2_CLASS));
+		/*
+		 * Class with multiple interfaces have an Interface
+		 * Association Descriptor available, use an appropriate triple
+		 * to indicate it.
+		 */
+		usbd_device_set_code_triple(usbd, speed, USB_BCC_MISCELLANEOUS, 0x02, 0x01);
+	} else {
+		usbd_device_set_code_triple(usbd, speed, 0, 0, 0);
+	}
+
+	return 0;
+}
+
+static struct usbd_contex *usb_init_next_usbd_init(void)
+{
+	int err;
+
+	USBD_DEVICE_DEFINE(usbd, DEVICE_DT_GET(DT_NODELABEL(usbd)),
```

Comment:
I wanted to keep all of the functions and macros that are specific to USB next stack separated from implementation of generic functionalities. Whole USB next configuration is currently kept together too. That's why I would prefer to keep it as is.
The `usbd_ctx` pointer is an exception here, because we need to access it if we want to trigger USB remote wakeup with USB next stack.

Reply:
I think the remote wakeup should rather be up to the stack, which is precisely what https://github.com/zephyrproject-rtos/zephyr/pull/73362 does.

Reply:
> I wanted to keep all of the functions and macros that are specific to USB next stack separated from implementation of generic functionalities. Whole USB next configuration is currently kept together too. That's why I would prefer to keep it as is. The `usbd_ctx` pointer is an exception here, because we need to access it if we want to trigger USB remote wakeup with USB next stack.

I don't understand why defining the instance in the same line as you define currently pointer to it would invalidate your assumptions

Reply:
> I think the remote wakeup should rather be up to the stack, which is precisely what [zephyrproject-rtos/zephyr#73362](https://github.com/zephyrproject-rtos/zephyr/pull/73362) does.

That may be a nice improvement. I will look into it more in my subsequent PRs.

Reply:
> > I wanted to keep all of the functions and macros that are specific to USB next stack separated from implementation of generic functionalities. Whole USB next configuration is currently kept together too. That's why I would prefer to keep it as is. The `usbd_ctx` pointer is an exception here, because we need to access it if we want to trigger USB remote wakeup with USB next stack.
> 
> I don't understand why defining the instance in the same line as you define currently pointer to it would invalidate your assumptions

It is "bundled" with assigning VID and PID. It prefer to assign all of the properties in one place if possible. Once improvement linked by @tmon-nordic goes in, we may be able to get rid of the global reference pointer.